ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotBlockConfirmer.java

```diff
@@ -178,6 +200,58 @@ private CompletableFuture<BlockHeader> executePivotQuery(final long blockNumber)
     return pivotHeaderFuture;
   }
 
+  private CompletableFuture<BlockHeader> executePivotQueryWithPeerTaskSystem(
+      final long blockNumber) {
+    GetHeadersFromPeerTask task =
+        new GetHeadersFromPeerTask(
+            blockNumber, 1, 0, GetHeadersFromPeerTask.Direction.FORWARD, protocolSchedule);
+    Optional<EthPeer> maybeEthPeer;
+    final EthPeer ethPeer;
+    try {
+      do {
+        if (isCancelled.get()) {
+          return CompletableFuture.failedFuture(
+              new InterruptedException("Pivot block confirmation has been cancelled"));
+        }
+        maybeEthPeer =
+            ethContext
+                .getEthPeers()
+                .getPeer(
+                    (p) ->
+                        task.getPeerRequirementFilter().test(p)
+                            && !peerIdsUsed.contains(p.nodeId()));
+      } while (maybeEthPeer.isEmpty() && waitForPeer());
+      ethPeer = maybeEthPeer.get();
+      peerIdsUsed.add(ethPeer.nodeId());
+    } catch (InterruptedException e) {
+      return CompletableFuture.failedFuture(e);
+    }
+
+    return ethContext
+        .getScheduler()
+        .scheduleServiceTask(
+            () -> {
+              PeerTaskExecutorResult<List<BlockHeader>> taskResult =
+                  peerTaskExecutor.executeAgainstPeer(task, ethPeer);
+              if (taskResult.responseCode() == PeerTaskExecutorResponseCode.INTERNAL_SERVER_ERROR) {
+                // something is probably wrong with the request, so we won't retry as below
+                return CompletableFuture.failedFuture(
+                    new RuntimeException("Unexpected internal issue"));
+              } else if (taskResult.responseCode() != PeerTaskExecutorResponseCode.SUCCESS
+                  || taskResult.result().isEmpty()) {
+                // recursively call executePivotQueryWithPeerTaskSystem to retry with a different
+                // peer.
+                return executePivotQueryWithPeerTaskSystem(blockNumber);
+              }
+              return CompletableFuture.completedFuture(taskResult.result().get().getFirst());
+            });
+  }
+
+  private boolean waitForPeer() throws InterruptedException {
```

Comment:
What thread does this block? I think the `confirmPivotBlock` is called before the sync pipeline, so this might not be on an executor thread.

Reply:
You're correct this will block a thread which we probably don't want to be blocked. However, this matches the existing behavior of the old system. This is definitely something we could and should clean up later.

Reply:
Might be reading the code wrong but I thought the current behaviour used `WaitForPeerTask` which looks like it is done async and will return as soon as we have a new peer and timing out after 5 seconds.